### ISSUE TITLE
Bug in ...\org.dita.pdf2\xsl\common\topicmergeImpl.xsl - Reference to glossary entry does not work

### ISSUE THREAD

**Issue by hscherzer** [dita-ot/dita-ot#2244]
Test: my source is like ...

```xml
in the example to associated <xref keyref="gls/gls_oxygen"/> to the file ...
```

whereas _glossary.dita_ in the ditamap has `keys=gls`
The glossary entry in _glossary.dita_ is like

```xml
 <glossentry id="gle_oxygen">
      <glossterm id="gls_oxygen">Oxygen</glossterm>
      <glossdef id="glc_oxygen">oxygen XML editor</glossdef>
   </glossentry>
```

The xref in oxygen of course resolves correct.
However, the final PDF contains
"As **#unique_5/unique_5_Connect_42_gls_oxygen** stores much"
which apparently is a proof that the link was not solved.
I found the solution in `\org.dita.pdf2\xsl\common\topicmergIimpl.xsl` which has to be applied in the official distribution ...

Version 2.2.2 line 159

```xml
<xsl:template match="*[contains(@class,' topic/topic ')] | dita-merge/dita">
```

to be changed into

```xml
<!--HSX I had to exclude glossentry from the match in order to 
        propagate the correct id prefix. Other templates do this correctly.
        The situation occurred when trying to refer xref a glossterm -->
    <xsl:template match="*[contains(@class, ' topic/topic ')]
                          [not(contains(@class, ' glossentry/glossentry '))] |
                          dita-merge/dita">
````

The fix works since more than 6 months will all versions since.
See the example PDF files on page 4
[topicmergeFix_Scherzer.zip](https://github.com/dita-ot/dita-ot/files/153061/topicmergeFix_Scherzer.zip)

**Comment by raducoravu** [comment]
How is the "glossary.dita" referenced in the DITA Map? If it is referenced using "glossref", it needs to have the attribute "print="yes"" set to it, see: https://github.com/dita-ot/dita-ot/issues/2173

**Comment by xephon2** [comment]
@hscherzer : Your usage of `<xref>` violates the spec:

> When creating a cross-reference, link to the element structure, not the `<title>` element of the object. For example, to create a cross-reference to a figure, link to the `<fig>` element, not the `<title>` element within the `<fig>` element. Output processing determines whether the text of the `<title>` element is used to render the cross-reference.

See [DITA 1.3 > xref](http://docs.oasis-open.org/dita/dita/v1.3/os/part3-all-inclusive/langRef/base/xref.html#xref)

Have you tried the `<abbreviated-form>` element, see [DITA 1.3 > abbreviated-form](http://docs.oasis-open.org/dita/dita/v1.3/os/part3-all-inclusive/langRef/technicalContent/abbreviated-form.html#abbreviated-form)?

**Comment by hscherzer** [comment]
Thanks very much for your contributions ... I have investigated your comments about their impact ... here is my result

**1. Radu:** _I am not using glossref_ and I do understand the background of _print=yes_. Your answer is correct, yet ... it has little match with my actual finding/bug report. But I understand how you came to it.

**2. Stefan:** I know what you mean when you recommend to **xref** to e.g. **fig** instead of its _title_. However, you made a transfer from my _xref_ to a _glossterm_ into a true statement regarding recommended _xref_ to topics like _concept/fig/table_ ... but my case is different. 

BTW: If you _would_ refer to a _title_, this is not very wise but it is also **not** "violating the rules". The DitaSpec does not forbid this, it recommends in a note to do so, yet the DITA-OT looks after a potential _title_ if it hits an _xref_ on its parent. The specification (your reference) shows examples where you refer to very much inside topics e.g. paragraph.

But as said ... I know here you come from. You argue that _glossentry_ would be the main topic to be referred and I should not xref _glossterm_.

Yet - I'm sorry .. your transfer doesn't apply. Here is the reason.....

Many people (and your assumption) come from the idea that a glossary with _42 terms_ is made of _42 files_ each of them being a _glossary_ type and the parent element of those is "glossentry". For several good reasons this approach is unacceptable for me (and my company).

The most obvious one is that you really have a hard time to administrate so many little files and my glossaries have more than 140 entries. Others reasons are in the scope of glossary creation by parsing all xrefs in my ditamap and creating the relevant 42 entries from a master glossary with more than 1000 entries. (Works very efficiently !) Therefore I need another (allowed!) approach:

DITA offers the nice alternative of a **glossgroup - glosssentry - glossterm:glossdef** ... all in one file.

**Here is where the bug comes in:**

The DITA-OT is made to avoid the same ID on main topics (concept/task/reference/**glossentry**) by accident or double-appearance of the same chapter in a ditamap (allowed). 

To do so it _changes the id_ of any such element to a format like **#unique_N/unique_N_Connect_42_<original ID>**. For the main elements (concept etc.) the original ID is totally replaced by the simple "**#unique_N**" tag. Some post-work is done in "_topicmergeImpl.xsl_" but the actual assignment (I suppose) is done in the JAVA code which is harder to fix.

Usually xref-statements to those changed ids are correctly updated by the DITA-OT process.

For the case with the 42 files (your potential assumption) .... **glossentry** is the a top level thing and such update is relevant. Also your comment.

However - in my case - _glossentry_ is a child of _glossgroup_ (and that's legal). What happens without my suggested fix is ... the _topicmerge_ process produces new IDs for the glossentry because it does not distinguish between the case "under glossgroup" or "main topic in a glossary type file". However the XREF statements to the entry are not updated. That's why the link does not work. (It IS printed ... but the link does not resolve).

I fixed the problem by not allowing the DITA-OT to update _glossentry_ if its parent is _glossgroup_. This works for both approaches.

However - after your precious comments ... I had to improve the suggested fix to make both approaches work:
**org.dita.pdf2\xsl\common\topicmergImpl.xsl** about Line 159

`    <xsl:template match="
            *[contains(@class, ' topic/topic ')]
            [not(contains(../@class, ' glossgroup/glossgroup '))]
            | dita-merge/dita">`

would be the right test. You can see that I do not want to reassign children of glossgroup which finally saves my XREF. It works for the glossterm/glossdef but glossentry is already trashed prior the match because the JAVA code already changed it.

This is also the reason why I don't refer to the 'main' tag (e.g. glossentry id=gle_<word>) and I am using reference to glossterm id = gls_<term>, from the viewpoint of my _xref_ ... the glossentry id is already 'too much improved' and the Xref was not updated to follow. But glossterm/glossdef have survived and I can bring them to correct behavior by the suggested bug fix.

With my fix ... all of that works perfectly and has no impact on the existing approaches. That's why I have to reference the _glossterm_ instead of the _glossentry_ (which cannot be referred to since it's id is 'improved').

I regret the long explanation but it is no less complicated as that. Yet the fix is small and you are invited to test the small file that I had attached to verify.

Thanks again for your contributions. **Do I have to fork / commit / push to get the change on the GIT-HUB or is the ISSUE-method the right way to suggest the fix?**

Thanks again and Cheers ... Helmut

**Comment by hscherzer** [comment]
In the above comment there is a correct statement that you cannot refer to a glossentry under a glossgroup because it's id is changed (bug, originally in dost.jar) since the DITA-OT does not distinguish between the collected glossentries (under glossgroup) and those who are the root of a file (separate files approach).
This can be fixed and I attached the associated files. The fix has to be paid with the following statement in **topicmergeImpl.xsl:285**

    <xsl:template match="@*" priority="-1">
        <xsl:choose>
            <xsl:when test="(name() ='id') and
                            contains(../@class, ' glossentry/glossentry ') and
                            contains(../../@class, 'glossgroup/glossgroup ')">
                <xsl:attribute name="id" select="concat(../../@id, '_Connect_42_', ../@oid)"/>
            </xsl:when>
            <xsl:otherwise>
                <xsl:copy-of select="."/>
            </xsl:otherwise>
        </xsl:choose>
    </xsl:template>
`
This fix reconstructs the correct id of the glossentry if it lives under a glossgroup.

To summarize .. the other fix in **topicmergImpl.xsl:160** was


    <xsl:template match="
            *[contains(@class, ' topic/topic ')]
            [not(contains(../@class, ' glossgroup/glossgroup '))]
            | dita-merge/dita">
        <xsl:param name="parentId"/>

To support the use of **abbreviated-form** there is another fix required in **abbrev-domain.xsl:22**

<xsl:template match="*[contains(@class,' abbrev-d/abbreviated-form ')]" name="topic.abbreviated-form">
    <xsl:variable name="keys" select="@keyref"/>
    <xsl:variable name="target" select="key('id', substring(@href, 2))"/>
    
    <xsl:variable name="gletarget" 
        select="key('id', substring(substring-before(@href, '/'), 2))/glossentry[@id = substring-after(current()/@href, '/')]"/>
    
    <xsl:choose>
      <xsl:when test="$keys and $gletarget/self::*[contains(@class,' glossentry/glossentry ')]">
          <xsl:call-template name="topic.term">
          <xsl:with-param name="contents">
            <xsl:variable name="use-abbreviated-form" as="xs:boolean">
                <xsl:apply-templates select="." mode="use-abbreviated-form"/>
            </xsl:variable>
            <xsl:choose>
              <xsl:when test="$use-abbreviated-form">
                  <xsl:apply-templates select="$gletarget" mode="getMatchingAcronym"/>
              </xsl:when>
              <xsl:otherwise>
                <xsl:apply-templates select="$gletarget" mode="getMatchingSurfaceForm"/>
              </xsl:otherwise>
            </xsl:choose>
          </xsl:with-param>
        </xsl:call-template>
      </xsl:when>
      <xsl:when test="$keys and $target/self::*[contains(@class,' glossentry/glossentry ')]">
        <xsl:call-template name="topic.term">
          <xsl:with-param name="contents">
            <xsl:variable name="use-abbreviated-form" as="xs:boolean">


I attached the fixes, the associated DITA source and the PDF output to show the bug and how the suggested fix works.

[FixesTest.zip](https://github.com/dita-ot/dita-ot/files/160220/FixesTest.zip)

**Comment by shudson310** [comment]
I can verify that I had the same problem as was reported. I am also pleased to report that hscherzer's fixes seem to have corrected the problem. I would also recommend that these fixes be included in the base.

**Comment by raducoravu** [comment]
We have another user with the same problem.
So the topic has an abbreviated form like:

       <abbreviated-form keyref="glossgroup/gloss_2G"/>

the DITA Map defines the "glossgroup" key like:

      	<topicref href="TestGlossGroup.dita" keys="glossgroup"/> 

where the "TestGlossGroup.dita" has content like:

            <glossgroup id="glossgroup">
            ................
              <glossentry id="gloss_2G">

but I'm not sure if the DITA specs would allow this to work. The "gloss_2G" part in the keyref should point be the ID of a DITA element and not to a DITA "glossentry" which is a topic type.

**Comment by hscherzer** [comment]
**Short version of my comment:** With my provided fix ... it will work ... and it is not forbidden by the Dita Spec (1.3).

**Long Version:**
As mentioned earlier in this thread ... I checked the Dita Spec (1.3) for the idea 'reference the element' and it does **not** provide _any constraint_ to referencing anything that has an id. There is an understandable _recommendation_ to reference the 'element' if it has a _title_, valid for concept/task/reference/topic/table/figure etc. but that's not applicable for the current case.

For the **glossgroup** ... The claim to reference the 'element' can only mean to get to the HIGHEST **identifiable** piece of  the structure. In a glossgroup - that is **glossentry**. If the DITA spec would forbid to reference that, they shall forbid the use of glossgroup (because then it wouldn't make any sense, since you can no more reference to your precious glossary entries).

This thread does not even suggest the case as a 'problem' - it provides the fix (or let's say 'extension') to the Dita-OT.

With the fix ... all combinations (see above) do work very well and the examples attached demonstrate the effectiveness. You can then reference the **glossentry** or the **glossterm** and the content of the glossterm will show up in the _xref-linktext_ unless you have explicitly written the linktext into the xref.

So feel free to try the fix yourself or to suggest it to the core team, I don't know how to get fixes brought into the official versions - to me the core authors are a bit of a closed group, but I might miss the point because I simply haven't found the path to 'contribute'. Maybe that's only a lack technical knowledge.

**Comment by robander** [comment]
I've been reading through the discussion here, and trying to understand the root issue, which I've had a hard time pulling out. @hscherzer this is partly based on your comment that this would be better fixed in `dost.jar` -- any time we're going to fix something, I'd prefer that we fix it in the right spot.

I started by trying to understand this issue in as general a way as possible. I set up a simple test to try and figure out what is working out of the box today. I used a simple map (not a bookmap), with a topic full of links. The link targets were all part of the map. The link targets are:
- A simple topic document. Link to the document (file name), to the root topic (file name + topic ID), to the title element of the root topic, to a figure in the topic, and to a title element inside the figure. Also link to a nested topic (file name + topic ID), title element of the nested topic, figure in the nested topic, and title element inside the figure.
- A simple reference document. Link to the document (file name), to the root reference element (file name + topic ID), to the title element of the reference topic, to a figure in the reference, and to the title element inside the figure. Also link to a nested reference topic (file name + topic ID), to the title element of the nested reference topic, to a figure in the nested reference, and to the title element inside the figure.
- A simple glossentry document. Link to the document (file name), to the root glossentry (file name + topic ID), to the title element of the glossentry topic, to a figure in the glossentry, and to a title element inside the figure
- A glossgroup document. Link to the document (file name), to the root glossgroup topic (file name + topic ID), and to the title element of the glossgroup. Also link to a nested glossentry topic (file name + topic ID), to a title within the glossentry, to a figure in the glossentry, and to the title of the figure. Also link to a second nested glossentry topic (file name + topic ID), and again to the title, the figure, and the figure title.

Using DITA-OT 2.2.3, *all* links resolve correctly in the PDF. That is, if I click on the link text, they go to the right spot).

The more general issue seems to be that whenever you link to a `<title>` element, the PDF output will use something ugly like `#unique_5/unique_5_Connect_42_gls_oxygen` as the link text. (HTML based output is also ugly, but it looks more like the original `@href` value). There is also an error message `DOTX023E` for each link to a `<title>` element. This happens for links to a topic title, nested topic title, reference title, glossentry title, glossgroup title, or figure title. 

From that - I think the more general issue here is that DITA-OT uses bad link text for any link that goes directly to a `<title>` element. [Why that is done, and whether that's a good idea in this case or in any case, is irrelevant to this issue.] I think that if links to titles used the title text, it would resolve this issue. If that's not the case, somebody please correct me...

**Comment by robander** [comment]
Attaching my zip of sample files, and the PDF using out-of-the-box DITA-OT 2.2.3.
[2224.zip](https://github.com/dita-ot/dita-ot/files/192444/2224.zip)

**Comment by robander** [comment]
@raducoravu about your comment: 
> `<abbreviated-form keyref="glossgroup/gloss_2G"/>`
...
> but I'm not sure if the DITA specs would allow this to work. The "gloss_2G" part in the keyref should point be the ID of a DITA element and not to a DITA "glossentry" which is a topic type.

That syntax should not work, for the reason you suggested. The `<glossentry>` element is a topic specialization; as with any topic, it can and should be linked to directly. A normal nested topic could look like this:
```xml
<topic id="root">
  <title>Root topic</topic>
  <topic id="nest">
    <title>Nested topic</topic>
  </topic>
</topic>
```

In that case, you would link to the nested topic using `file.dita#nest`. The syntax `file.dita#root/nest` is invalid, because that would be a link to a non-topic element inside of the topic with `id="root"`.

The same is true for a `<glossentry>` element, which is a specialized topic. I may have the following:
```xml
<glossgroup id="group">
  <title>My group of entries</title>
  <glossentry id="a"><glossterm>A term</glossterm></glossentry>
  <glossentry id="b"><glossterm>B term</glossterm></glossentry>
  <glossentry id="c"><glossterm>C term</glossterm></glossentry>
  <glossentry id="d"><glossterm>D term</glossterm></glossentry>
</glossgroup>
```

In that case the proper way to link to "C term" (without keys) would be `href="file.dita#c"` (not `href="file.dita#group/c"`, which would again mean a non-topic element with `id="c"` inside of the topic element with `id="group"`).

The syntax `keyref="glossgroup/c"` (where the key `glossgroup` is associated with the document) is not valid for the same reason that `href="file.dita#group/c"` is not valid. In fact it would translate to the same thing -- a non-topic element with `id="c"' inside of the root topic.

If you want to set up a key to link to "c", or any other glossentry, that can be done in the map -- because a key can be associated directly with any topic or sub-topic:
```xml
<keydef keys="group" href="file.dita#group"/> <!-- "group" = the root group -->
<keydef keys="term-a" href="file.dita#a"/> <!-- "term-a" = the topic level element with id="a" -->
<keydef keys="term-b" href="file.dita#b"/> <!-- "term-b" = the topic level element with id="b" -->
```

Note -- I'm not arguing that this is obvious / easy / the fastest way things could have been authored. But it's a result of the fact that the `<glossentry>` element is a specialized topic, and the syntax for linking within DITA files + the syntax for `keys` / `keyref` is all based on the topic as a primary linkable object within a file.

**Comment by raducoravu** [comment]
@robander I understand + agree

**Comment by robander** [comment]
> Rob - I can follow your idea, but I'm afraid I am too much an engineer that I'd eat the philosophy argument when I need a practical solution. I would agree to the philosophy ("do not reference a specialized topic")

I'm afraid I really do not understand this. Perhaps I wrote too much and my point was lost. What I was trying to say:

- The issue you originally reported is not limited to glossary terms. You demonstrated that cross references to `<glossterm>` give bad output. The `<glossterm>` element is specialized from `<title>`. I pointed out that this is a general issue with all cross references to titles, including topic titles, figure titles, `<glossterm>`, and so on.
- I suggested a practical fix: if you cross reference any title, the generated link text from DITA-OT should be the title text. This will fix cross references to _all_ titles, including any reference to the `<glossterm>` element. There is a DITA-OT module dedicated to retrieving link text which is the obvious place for this fix.
- I never said "do not reference a specialized topic". In fact, I suggested the opposite. References to specialized topics (in this case, `<glossentry>`) are appropriate, proper, and encouraged. Link text should (and does) resolve to the term. Links should (and do) take you directly to the entry.

**Comment by robander** [comment]
You also mentioned a separate issue: 
> I need to refer to a DITA.PDF document from externally - which means to create named destinations - which the DITA-OT does from the id ... however by the allocation of unique_connect_42 prefix, the actual destinations are unpredictable, which is a problem. I have a special trick to get around that ...

I understand that this is a problem. I would be happy to have a solution for it. But, that too is a much broader issue than glossary terms. I don't know if your solution to that issue addresses the broader issue. I know the DITA-OT team would rather discuss it independently.

Overall, I've seen / written too much code over the years using special tricks that seemed good at the time. Special tricks that fix one narrow issue now very often result in expensive updates later. In a DITA-OT context, special tricks and fixes/features that were implemented in the wrong location accounted for much of the technical debt we cleaned up in DITA-OT 2.0.

**Comment by hscherzer** [comment]
Rob, thanks very much for your fast and helpful answer. I will need some time to tryout your valuable example and to react more precise on your reflections! Already now I can say that I agree to the following points
- Yes ... too many well meant fixes can entirely spoil the idea of the Dita-OT and I agree that therefore some conservative approach to changes is a life saver
- Yes ... the discussion on the named destinations is not to be done here. The problem is solved, yet it should go to another thread.

As said - I will check your good examples to see how the discussion might converge.
Cheers ... Helmut

**Comment by robander** [comment]
Jarno added a fix in #2283 that will fix link text for any link to any `<title>` element (including the specialized title element `<glossterm>`). That fix resolves the issue originally reported here about bad link text when `<glossterm>` is the target of a cross reference. The fix is part of the code module that resolves all generated link text, meaning that it also works for other formats like HTML, which also used bad text previously. [Also, when creating this fix, Jarno noticed I used the wrong ID in one of my test case documents, so beware if using those attached above.]

The fix is already checked in to the `develop` branch for the next major release, and is also checked in to the `hotfix/2.2.4` branch; the 2.2.4 patch release should be out within two weeks.

**Comment by hscherzer** [comment]
Great news ! - I am eager to see the fix and thanks to all contributors (in particular Rob) for caring and the good discussion... Cheers ... Helmut
